backend/log/context_grpc.go

```diff
@@ -0,0 +1,55 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	logParamsCtxMetadataKey = "loggerParamsCtxMetadata"
+	logParamSeparator       = ":"
+)
+
+// WithContextualAttributesForOutgoingContext will append the given key/value log parameters to the outgoing context.
+func WithContextualAttributesForOutgoingContext(ctx context.Context, logParams []any) context.Context {
+	if len(logParams) == 0 || len(logParams)%2 != 0 {
+		return ctx
+	}
+
+	for i := 0; i < len(logParams); i += 2 {
+		k := logParams[i].(string)
+		v := logParams[i+1].(string)
+		if k == "" || v == "" {
+			continue
+		}
+
+		ctx = metadata.AppendToOutgoingContext(ctx, logParamsCtxMetadataKey, logParam(k, v))
+	}
+
+	return ctx
+}
+
+// ContextualAttributesFromIncomingContext returns the contextual key/value log parameters from the given incoming context.
+func ContextualAttributesFromIncomingContext(ctx context.Context) []any {
+	logParams := metadata.ValueFromIncomingContext(ctx, logParamsCtxMetadataKey)
+	if len(logParams) == 0 {
+		return nil
+	}
+
+	var attrs []any
+	for _, param := range logParams {
+		kv := strings.Split(param, logParamSeparator)
```

Comment:
it may be needed to scape the logParamSeparator in case a param includes `:` on it?

Reply:
Great catch. I just updated it to be 3 characters instead 👀 rather than making things more complex. WDYT?

Reply:
I guess that's fine unless the parameter is a ipv6 😅 I was more thinking about replacing `:` with `\:` but it can be left for later.